imblearn/over_sampling/_smote_nc.py

```diff
@@ -0,0 +1,246 @@
+    >>> X_res, y_res = sm.fit_resample(X, y)
+    >>> print('Resampled dataset samples per class {}'.format(Counter(y_res)))
+    Resampled dataset samples per class Counter({0: 900, 1: 900})
+
+    """
+
+    def __init__(self,
+                 sampling_strategy='auto',
+                 random_state=None,
+                 k_neighbors=5,
+                 m_neighbors='deprecated',
+                 out_step='deprecated',
+                 kind='deprecated',
+                 svm_estimator='deprecated',
+                 n_jobs=1,
+                 ratio=None,
+                 categorical_features=None):
+        super(SMOTENC, self).__init__(sampling_strategy=sampling_strategy,
+                                      random_state=random_state,
+                                      k_neighbors=k_neighbors,
+                                      m_neighbors=m_neighbors,
+                                      out_step=out_step,
+                                      kind=kind,
+                                      svm_estimator=svm_estimator,
+                                      n_jobs=n_jobs,
+                                      ratio=ratio)
+        self.categorical_features = categorical_features
+
+    def _fit_resample(self, X, y):
+        if self.categorical_features is None:
+            warnings.warn('No "categorical_features" were specified when '
+                          'this instance was created. Will fall back '
+                          'to normal SMOTE', RuntimeWarning)
+            return super(SMOTENC, self)._fit_resample(X, y)
+
+        feature_indices = check_array(self.categorical_features, ensure_2d=False,
+                                      ensure_min_samples=2, estimator=self)
+        n_features = X.shape[1]
+
+        if np.any(feature_indices > n_features):
+            raise ValueError('Indices of categorical features have to be less '
+                             'than number of features in X: X.shape=(%s, %s)'
+                             % X.shape)
+
+        self.categorical_feature_indices_ = feature_indices
+        continuous_cols = np.delete(np.arange(n_features),
+                                    np.arange(feature_indices[0],
+                                              feature_indices[-1]))
+        if len(continuous_cols) == 0:
+            raise ValueError('Looks like all features in X are '
+                             'categorical which is not supported. '
+                             'For this method to work X should have '
+                             'at least 1 continuous feature.')
+
+        if sparse.issparse(X):
+            # let StandardScaler handle per-column variance in sparse matrix
+            scaler = StandardScaler(with_mean=False,
+                                    with_std=True,
+                                    copy=False)
+            scaler.fit(X.tocsc()[:, continuous_cols])
```

Comment:
I thought that some types of sparse matrices (coo, dok) do not support column slicing? (it is applied before scaler.fit() is called.)

Reply:
oh you don't need to convert here.
We are making a sanity check in `fit_resample` in `base.py` to be sure that you will have only `csr` or `csc` matrices pass otherwise `check_array` will make the conversion:

https://github.com/scikit-learn-contrib/imbalanced-learn/blob/master/imblearn/base.py#L140